File: csharp/demo/demo08-return-value.cs
/*
 * GETTING RETURNED VALUES FROM THREADS
 */
using System;
using System.Threading;



class Demo08 : IRunnable
{
    public void run()
    {
        int[] result = new int[2];

        var thFoo = new Thread(() => result[0] = doubleValue(5));
        var thBar = new Thread(() => result[1] = doubleValue(80));

        thFoo.Start();
        thBar.Start();

        // Wait until thFoo and thBar finish
        thFoo.Join();
        thBar.Join();

        Console.WriteLine(result[0]);
        Console.WriteLine(result[1]);
    }


    private int doubleValue(int value)
    {
        return value * 2;
    }
}

File: cpp/cpp-boost/demo08-return-value.cpp
/*
GETTING RETURNED VALUES FROM THREADS
Using pointers or references (traditional way)
*/


#include <iostream>
#include <boost/ref.hpp>
#include <boost/thread.hpp>
using namespace std;



void doubleValue(int arg, int* res) {
    (*res) = arg * 2;
}



void squareValue(int arg, int& res) {
    res = arg * arg;
}



int main() {
    int result[3];

    boost::thread thFoo(&doubleValue, 5, &result[0]);
    boost::thread thBar(&doubleValue, 80, &result[1]);
    boost::thread thEgg(&squareValue, 7, boost::ref(result[2]));

    thFoo.join();
    thBar.join();
    thEgg.join();

    cout << result[0] << endl;
    cout << result[1] << endl;
    cout << result[2] << endl;

    return 0;
}
